nucliadb_vectors/src/fst_index/mod.rs

```diff
@@ -0,0 +1,327 @@
+//
+// If we do this in memory it's going to cost ~350MiB per 100k labels.
+// If we want less memory pressure, we could store labels in several FSTs and merge
+// search results.
+//
+use std::collections::HashSet;
+use std::fs::{File, OpenOptions};
+use std::io::{self, BufReader, BufWriter, Seek, Write};
+use std::path::{Path, PathBuf};
+use std::sync::RwLock;
+
+use bincode::serialize_into;
+use fst::automaton::{Automaton, Str};
+use fst::{IntoStreamer, Map, MapBuilder, Streamer};
+use memmap2::Mmap;
+use serde::{Deserialize, Serialize};
+
+use crate::VectorR;
+
+/// Struct that contains the label and a sorted list of node_addresses
+#[derive(Serialize, Deserialize, PartialEq, Debug, Clone)]
+pub struct Label {
+    pub key: String,
+    pub node_addresses: Vec<u64>,
+}
+
+pub struct LabelIndex {
+    pub fst_file_path: PathBuf,
+    pub records_file_path: PathBuf,
+    records_file: RwLock<BufReader<File>>,
+    fst_map: Map<Mmap>,
+}
+
+/// LabelIndex manages an FST file along with an index file.
+/// They are created by iterating on Label objects.
+impl LabelIndex {
+    const LABELS_FST: &str = "labels.fst";
+    const LABELS_IDX: &str = "labels.idx";
+
+    /// Returns true if FST files are present
+    pub fn exists(path: &Path) -> bool {
+        let records_file_path = path.join(Self::LABELS_IDX);
+        let fst_file_path = path.join(Self::LABELS_FST);
+        records_file_path.exists() && fst_file_path.exists()
+    }
+
+    /// Opens the index for read access.
+    /// `path` is the directory where the FST and index files are located.
+    pub fn open(path: &Path) -> VectorR<Self> {
+        let records_file_path = path.join(Self::LABELS_IDX);
+        let fst_file_path = path.join(Self::LABELS_FST);
+
+        let records_file = BufReader::new(File::open(&records_file_path)?);
+        let mmap = unsafe { Mmap::map(&File::open(&fst_file_path)?)? };
+        let fst_map = Map::new(mmap)?;
+
+        Ok(Self {
+            fst_file_path,
+            records_file_path,
+            fst_map,
+            records_file: RwLock::new(records_file),
+        })
+    }
+
+    /// Given a list of labels, return the intersection of all node addresses
+    pub fn get_nodes(&self, labels: &[String]) -> VectorR<HashSet<u64>> {
+        let mut records = self.records_file.write().unwrap();
+        let mut results: HashSet<u64> = HashSet::new();
+
+        for (index, label) in labels.iter().enumerate() {
+            if let Some(offset) = self.fst_map.get(label) {
+                let _ = records.seek(io::SeekFrom::Start(offset))?;
+                let label_struct: Label = bincode::deserialize_from(&mut *records)?;
+                let set: HashSet<u64> = label_struct.node_addresses.iter().cloned().collect();
+                if index == 0 {
+                    results.extend(&set);
+                } else {
+                    results = results.intersection(&set).cloned().collect();
+                }
+            }
```

Comment:
If a label is not in the map shouldn't the intersection be empty?

Reply:
good catch! fixed (+one test)